Src/TransmitData.c

```diff
@@ -53,29 +97,54 @@ void transmitImuData(AllData* data)
         osMutexRelease(data->accelGyroMagnetismData_->mutex_);
     }
 
-    uint8_t buffer[IMU_SERIAL_MSG_SIZE] = {0};
-
-    buffer[0] = IMU_HEADER_BYTE;
-    buffer[1] = IMU_HEADER_BYTE;
-    buffer[2] = IMU_HEADER_BYTE;
-    buffer[3] = IMU_HEADER_BYTE;
-    writeInt32ToArray(&buffer, 4, accelX);
-    writeInt32ToArray(&buffer, 8, accelY);
-    writeInt32ToArray(&buffer, 12, accelZ);
-    writeInt32ToArray(&buffer, 16, gyroX);
-    writeInt32ToArray(&buffer, 20, gyroY);
-    writeInt32ToArray(&buffer, 24, gyroZ);
-    writeInt32ToArray(&buffer, 28, magnetoX);
-    writeInt32ToArray(&buffer, 32, magnetoY);
-    writeInt32ToArray(&buffer, 36, magnetoZ);
-    buffer[IMU_SERIAL_MSG_SIZE - 1] = 0x00;
+    //construct the message in the format accelXYZ, gyroXYZ, MagnetoXYZ
+    uint8_t message[IMU_SERIAL_MSG_SIZE] = { 0 };
+    int messageindex = 0;
+    message[0] = IMU_HEADER_BYTE;
+    messageindex++;
+    writeInt32ToArray(message, messageindex, accelX);
+    messageindex += 4;
+    writeInt32ToArray(message, messageindex, accelY);
+    messageindex += 4;
+    writeInt32ToArray(message, messageindex, accelZ);
+    messageindex += 4;
+    writeInt32ToArray(message, messageindex, gyroX);
+    messageindex += 4;
+    writeInt32ToArray(message, messageindex, gyroY);
+    messageindex += 4;
+    writeInt32ToArray(message, messageindex, gyroZ);
+    messageindex += 4;
+    writeInt32ToArray(message, messageindex, magnetoX);
+    messageindex += 4;
+    writeInt32ToArray(message, messageindex, magnetoY);
+    messageindex += 4;
+    writeInt32ToArray(message, messageindex, magnetoZ);
+    messageindex += 4;
+    //Find the final length of the encoded buffer based on number of overlaps
+    int encoded_message_length = IMU_SERIAL_MSG_SIZE;
+
+    for (int i = 0; i < IMU_SERIAL_MSG_SIZE; i++)
+    {
+        //If byte is F0 or F1, requires two bytes to represent(F1F2 or F1F3), so one additional byte is needed every time F0 or F1 is encountered
+        if (message[i] == F0_CHAR1 || message[i] == F1_ESCAPE)
```

Comment:
I also noticed that F0_CHAR1 and F1_ESCAPE are both equal to 0xF1. Change F0_CHAR1 to F0_ESCAPE.

Reply:
I think that F0_CHAR1 and F0_CHAR2 are kind of confusing so I renamed them to F0_REPLACEMENT_1 and F0_REPLACEMENT_2. I kept all the macros in case we ever need to change the flags and escapes.

Reply:
That's a good idea!